students/allen_maxwell/Lesson_09/donor_models.py

```diff
@@ -0,0 +1,152 @@
+
+    @property
+    def name(self):
+        '''Returns donor name'''
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        '''Blocks setting name of the donor.'''
+        raise AttributeError('Name cannot be set')
+
+    @property
+    def donations(self):
+        '''Returns the donation data.'''
+        return self._donations
+
+    @donations.setter
+    def donations(self, donation):
+        '''Blocks setting donations for the donor.'''
+        raise AttributeError('Donations cannot be set')
+
+    @property
+    def total_donations(self):
+        '''Returns total donation.'''
+        return self._donations[0]
+
+    @staticmethod
+    def add_total_donations(self, donation):
+        '''Adds the new donation value to the current total donations.'''
+        self._donations[0] = self.total_donations + self.check_donation_value(donation)
+
+    @staticmethod
+    def check_donation_value(donation):
+        '''Checks if the donation is greater than zero.'''
+        if donation > 0:
+            return donation
+        else:
+            raise AttributeError('Donations must be positive values, greater than zero')
+
+    @property
+    def num_donations(self):
+        '''Returns number donations.'''
+        return self._donations[1]
+
+    @staticmethod
+    def add_num_donations(self):
+        '''Adds one to the number of donations.'''
+        self._donations[1] = self.num_donations + 1
+
+    @property
+    def avg_donations(self):
+        '''Calculates the average donations of the donor.'''
+        return self.total_donations / self.num_donations
+
+    @avg_donations.setter
+    def avg_donations(self, value):
+        '''Blocks setting average donations for the donor.'''
+        raise AttributeError('Average donation cannot be set')
+
+    def new_donation(self, donation):
+        '''Adds a new donation to the donors donation data.'''
+        self.add_total_donations(self, donation)
+        self.add_num_donations(self)
+
+    def save_donor_file(self, letter):
+        '''Saves a letter for the donor to file.'''
+        first_name = self.name.split()[0]
+        last_name = self.name.split()[1].strip(',')
+        file_path = os.path.join("./{}_{}.txt".format(first_name, last_name))
+        with open(file_path, 'w+') as new_file:
+            new_file.write(letter)
+
+
+def sort_key(values):
+    '''Sorts the DonorCollection by historical value.'''
+    return values[1]
+
+
+class DonorCollection(Donor):
+    '''Collection of donors.'''
+
+    def __init__(self, *args):
+        self._donors = {donor.name: donor for donor in args}
+
+    @property
+    def donors(self):
+        '''Returns a collection of donors.'''
+        return self._donors
+
+    def is_donor_present(self, donor_name):
+        '''Checks if the donor is present in the collection.'''
+        for donor in self.donors.values():
+            if donor_name == donor.name:
+                return True
```

Comment:
why would you loop through the dictionary? the key is the name? how could you utilize a dictionary here?

Reply:
Updated changes in Lesson 9 Assignment Update